### DISCUSSION TITLE
Installation error, runnable.py missing

### DISCUSSION THREAD

**Discussion by liuyanulm** [Ulm-IQO/qudi-core#30] [Q&A] [answered]
runnable.py missing from `qudi-iqo-modules`.

while "Main Window" does pop out, when `qudi` commands sent in the `qudi-envs` in the Anaconda prompt, along with some error reporting.

![image](https://user-images.githubusercontent.com/31623230/188093081-98913fbf-a4e3-4147-931f-13a775703d70.png)

installed in a ThinkStation PC, Win 10 OS, with latest Anaconda Python 3.9.

qudi-core installation:
        step 1, create Python3.9 env via Variant 2 by Anaconda,
        step 2, activate env via Variant 2
        step 3, via Variant 1 installing from PyPI

qudi-iqo-modules installation followed the official instructions step by step.

**Comment by liuyanulm** [comment, marked as answer]
The reported error associated with `qudi-kernel` is solved, by running `qudi-install-kernel` command in the envs directroy.
while `runnable.py` still not found.

**Reply by timoML** [reply]
The solution you described is Step 3 of the [qudi-core installation guide](https://ulm-iqo.github.io/qudi-core/setup/installation.html). Maybe we can try to make this step more apparent to the reader.

Concerning the runnable.py. As the "qudi" command runs from console, I strongly suspect that the file is there, but at a different location than you're looking. As you installed the core from PyPi, can you post the files stored in the path returned from (mind activating your qudi venv) `python -c "import os, sys; print(os.path.dirname(sys.executable)+'\Lib\site-packages\qudi')"`?

**Reply by Neverhorst** [reply]
As @timoML already pointed out, the `runnable.py` is not missing it's just installed in the `site-packages` location of your Python environment. This is because the `qudi` application is entirely contained within `qudi-core`. `qudi-iqo-modules` is just installing additional module libraries to use in the `qudi-core` application.
When installing qudi from PyPI it does not really make sense to start it from an IDE like PyCharm because you can not alter the code anyway but just start it as it is. This is why you also do not need the `runnable.py` in this case. Just start qudi from command line by entering `qudi` (after activating the Python environment of course). You can also create a small batch script if you like that does these steps for you and then you just need to click this script to start qudi.

If you want to edit qudi source code you need to install it in development mode.

**Reply by timoML** [reply]
I think there is the use case when you install qudi-core from pyPi and qudi-iqo-modules (or other addons) in dev mode. In this case, starting runnable.py from IDE is a convenient feature. Of couse, altering qudi-core code is not possible then.

**Reply by Neverhorst** [reply]
Also in the case you described, you do not need `runnable.py`. PyCharm run configurations can also execute shell scripts. In other words, you can run the `qudi` executable command from within PyCharm:
![image](https://user-images.githubusercontent.com/18266223/188108604-cfa6adba-f355-41f8-80a0-281336363cc6.png)

![image](https://user-images.githubusercontent.com/18266223/188108840-26d95d31-2c25-4469-8089-80cf96474f9d.png)

**Reply by timoML** [reply]
Good point. Might be the easier approach for the iqo-modules installation guide...

**Reply by liuyanulm** [reply]
> python -c "import os, sys; print(os.path.dirname(sys.executable)+'\Lib\site-packages\qudi')"

Hi Timo,
here is what is reported after your script. So no files returned.
![image](https://user-images.githubusercontent.com/31623230/188258393-4c4e728d-a2e7-4735-b68b-a142e58e7607.png)

This goes to the python site-pacages directory, while qudi is installed under `qudi-core` or the qudi environment, and it was found there.

The `runnable.py` is found by reading through all comments above, under the directory of `qudi-core`

**Reply by liuyanulm** [reply]
This is where I found it.
![image](https://user-images.githubusercontent.com/31623230/188258728-07ac59a6-8454-4b4b-bfc3-02febba82a94.png)

**Reply by Neverhorst** [reply]
As I mentioned above, it does not really make sense to search for `runnable.py` in case of a PyPI install of `qudi-core`. Use the executable script `qudi` to run it. See my screenshots above to get a hint how to do that from PyCharm (the working directory does not matter and can be set to whatever you want). This is how it was meant to be used and you do not have to care about where the code is located on your machine.

**Reply by timoML** [reply]
I think we should indeed change the iqo-modules guide to follow @Neverhorst's solution.

@liuyanulm In addition, if you really need to locate the code in site-packages, you need to activate the correct venv before executing the posted command (like stated at the beginning of the instructions).

**Reply by Neverhorst** [reply]
Just to mention... using the `qudi` executable is actually the universal method of starting qudi no matter how it's installed. The `runnable.py` script is just there for historical reasons and to showcase e.g. the start of qudi from another python script.

**Reply by liuyanulm** [reply]
> Just to mention... using the `qudi` executable is actually the universal method of starting qudi no matter how it's installed. The `runnable.py` script is just there for historical reasons and to showcase e.g. the start of qudi from another python script.

@Neverhorst Don't get me wrong. I understand that I don't have to find where`runnable.py` is located. I raised this as I was a bit misled by the instructions, thought something must have gone wrong because this file was not shown as the example pic given in the installation guide.

**Reply by liuyanulm** [reply]
> I think we should indeed change the iqo-modules guide to follow @Neverhorst's solution.
> 
> @liuyanulm In addition, if you really need to locate the code in site-packages, you need to activate the correct venv before executing the posted command (like stated at the beginning of the instructions).

@timoML I agree with you to follow Niko's solution. and also change the installation guide.

**Reply by timoML** [reply]
Installation guide updated. Can you mark as answered @liuyanulm?

**Reply by liuyanulm** [reply]
Good job! Thank you @timoML
